src/actions/bridgeWriteContract.ts

```diff
@@ -0,0 +1,50 @@
+import {
+  Chain,
+  Transport,
+  encodeFunctionData,
+  WriteContractParameters,
+  Account,
+  Abi,
+  EncodeFunctionDataParameters,
+  WalletClient,
+} from 'viem'
+import { writeContract } from 'viem/actions'
+import { l1CrossDomainMessengerABI } from '@eth-optimism/contracts-ts'
+import { OpChainL2, OpChainL1 } from '@roninjin10/rollup-chains'
+
+export async function bridgeWriteContract<
+  TAbi extends Abi | readonly unknown[] = Abi,
+  TFunctionName extends string = string,
+  TChainL1 extends OpChainL1 | undefined = OpChainL1,
+  TChainL2 extends OpChainL2 | undefined = OpChainL2,
+  TAccount extends Account | undefined = Account | undefined,
+  TChainOverride extends Chain | undefined = Chain | undefined,
+>(
+  client: WalletClient<Transport, TChainL1>,
+  // TODO make this take an l2Chain that is decorated with l2 info such as the l1 contract addreses
+  { toChain, args, abi, address, functionName, ...restArgs }: { toChain: TChainL2 } & WriteContractParameters<TAbi, TFunctionName, TChainL1, TAccount, TChainOverride>,
```

Comment:
re `toChain` think we should allow for OpStack chains that possibly we haven't added yet?

Reply:
Absolutely! Easiest example is bunch of dev nets. If you have an l2 chain object whether a “known” chain such as ‘import {optimism} from roninjin10/rollup-chains or some custom chain you made following the rollup chain interface as long as you provide the config you can use the sdk.

this is also how the sdk works today. Today there are chains we added internally and also you can pass in your own contract addresses. This is an improvement on that in that the chain configs aren’t coupled to the library they are just an interface that extends wagmi chains

Reply:
Thanks! But how can they pass a different chain here? Just has to make the type confirm? Should we allow for just passing chain ID as an alt?

Reply:
Exactly right. the chain object contains the address for the messenger contract below so as long as we have a messenger contract everything is all good.